### PR TITLE
[wifi] generic wifi service and HAL

### PR DESCRIPTION
**Pull request by Victor-C-Zhang**
Create the structures necessary to implement and connect to wifi.
Create a wifi service that runs a busy-polling thread. Create callback functionality to detect wifi disconnection. Create HAL and stub impl.
TODO: use wifi service in antioch base and curl

### CODE REVIEW

File: connectivity/wifi/include/antioch/connectivity/wifi/wifi_connection_error.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+enum class ErrorReason {
+  ERROR_NO_SSID,
+  ERROR_OTHER,
+  ERROR_SOURCE_CODE,
+};
+
+std::string ErrorReason_name(ErrorReason reason);
+
+class WifiConnectionError : public std::runtime_error {
```
Comment (q349wang) [inline connectivity/wifi/include/antioch/connectivity/wifi/wifi_connection_error.h:19]: So by default exceptions are not supported on esp32. I think there is a way to enable them but I'll have to look into that, just a heads up

File: device/linux/wifi/wifi.cpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+wifi_err_t wifi_disconnect() {
+  return RES_OK;
+}
+
+bool wifi_is_disconnected() {
+  return true;
+}
+
+}  // namespace hal::wifi
```
Comment (q349wang) [inline device/linux/wifi/wifi.cpp]: Nit nl

File: device/linux/wifi/wifi.cmake [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# Copyright Antioch. All rights reserved.
+
+add_library(wifi-hal
+  ${CMAKE_CURRENT_LIST_DIR}/wifi.cpp
```
Comment (q349wang) [inline device/linux/wifi/wifi.cmake:4]: What's CMAKE_CURRENT_LIST_DIR do

File: connectivity/wifi/include/antioch/connectivity/wifi/wifi_service.h [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+   * Call this function before connect()
+   * @throws WifiConnectionError if unable to enable connectivity.
+   */
+  void start_connectivity();
+
+  /**
+   * Call this function after start_connectivity()
+   * @throws WifiConnectionError if unable to connect to the AP.
+   */
+  void connect();
```
Comment (q349wang) [inline connectivity/wifi/include/antioch/connectivity/wifi/wifi_service.h:36]: No disconnect?

File: device/common/wifi/wifi.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+wifi_err_t wifi_set_ssid(std::string ssid, std::string psk);
+wifi_err_t wifi_start_connectivity();
+wifi_err_t wifi_stop_connectivity();
+wifi_err_t wifi_connect();
+wifi_err_t wifi_disconnect();
+
+bool wifi_is_disconnected();
+
+}  // namespace wifi
+}  // namespace hal
```
Comment (q349wang) [inline device/common/wifi/wifi.h]: Nit nl

File: device/common/wifi/wifi.h [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+#pragma once
+
+#include <cstdint>
+#include <string>
+
+namespace hal {
+namespace wifi {
+
+enum wifi_err_t : int16_t {
```
Comment (q349wang) [inline device/common/wifi/wifi.h]: I prefer uint return codes personally as a stylistic choice

File: device/common/wifi/wifi.h [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  RES_NOOP = -1,
+  RES_OK = 0,
+};
+
+wifi_err_t wifi_hal_init();
```
Comment (q349wang) [inline device/common/wifi/wifi.h]: If you are supposed to check return codes for these, consider using `[[ nodiscard ]]`

**Review by q349wang — APPROVED** [review]
LGTM, just a few comments

File: device/linux/wifi/wifi.cmake (hunk 3 above)
Comment (Victor-C-Zhang) [reply device/linux/wifi/wifi.cmake:4]: it's the dir of the cmake file

File: connectivity/wifi/include/antioch/connectivity/wifi/wifi_service.h (hunk 4 above)
Comment (Victor-C-Zhang) [reply connectivity/wifi/include/antioch/connectivity/wifi/wifi_service.h:36]: no need

**Pull request merged**